traits/observation/_has_traits_helpers.py

```diff
@@ -170,3 +113,23 @@ def observer_change_handler(event, graph, handler, target, dispatcher):
             dispatcher=dispatcher,
             remove=False,
         )
+
+
+def ctrait_prevent_event(event):
+    """ Return true if the CTrait change event should be skipped.
+
+    Parameters
+    ----------
+    event : TraitChangeEvent
+    """
+    if event.old is Uninitialized:
+        return True
+
+    ctrait = event.object.traits()[event.name]
+    if (ctrait.type == TraitKind.trait.name
+            and ctrait.comparison_mode == ComparisonMode.equality):
+        try:
+            return event.old == event.new
+        except Exception:
```

Comment:
To keep the same semantics as in the original code, we probably want an extra `bool` call here: `return bool(event.old == event.new)`. The `bool` conversion is more likely to raise than the equality check - it should be very rare for an equality check to raise, but it's somewhat less rare for the interpretation of the result as a boolean to raise (e.g., NumPy arrays).

I think we should also be trying to get rid of this exception-swallowing behaviour at some point, but obviously not in this PR. Please could you open an issue?

(Same comments apply in `_change_accepted` below.)

Reply:
Sure I can open an issue, but actually I think it makes sense for the change handler to always fire if it fails to carry out an equality check....... maybe I should continue that discussion in that separate issue.

Reply:
Maybe. My concern is the usual one with catching `Exception`: that exception may well represent a coding error, and we don't want that coding error to be silently hidden. So I want to see that exception propagated in most cases (or at the very least, logged).

Reply:
Good point about the equality check could return something else that is not a bool. I changed that.

Reply:
Opened and referenced #1230